### ISSUE TITLE
#64 Incorrect calculation of "*" width elements 

### ISSUE THREAD

**Issue by void256** [nifty-gui/nifty-gui#92]
(original from https://sourceforge.net/p/nifty-gui/bugs/64/)

If you create a panel with width \* then it calculates its width without allowing for any margins set for other panels, this causes those other panels to then be pushed too far to the right.

For example:

```
panel(new PanelBuilder("ourTint") {
    {
        childLayoutHorizontal();
        width("100%");
        height("100%");

        panel(new PanelBuilder() {{
            width("*");
        }});                                            

        text() {{
            font("Interface/Fonts/Georgia_20_bold.fnt");
            textHAlignCenter();
            textVAlignCenter();
            color("#fddf");
            margin("6px");
            height("32px");
            width("150px");
        }});

        panel(new PanelBuilder() {{
            width("12px");
        }});
}});
```

The final 12 pixel wide panel is needed to correct for the problem as the 6 pixel margin on the left and right of the text is not included when the first panel calculates its width so the text gets pushed right too far right.

**Comment by philipwhiuk** [comment]
There's no margin attributes in the XML for 1.3.3 (using XSD: http://nifty-gui.sourceforge.net/nifty-1.3.xsd ). Is XML based stuff basically deprecated? If so, should I move to JavaBuilder type stuff (especially looking towards 1.4?)

In regard to this specific issue, probably should try and fit in with the DOM box-model behaviour.

**Comment by void256** [comment]
Yes, the problem is that I've not updated the XSD that is available online in a long time. So what you see online is an old version of the XSD. The current one that ships with 1.3.3 is correct and contains the margin attributes. This is the correct version: https://github.com/void256/nifty-gui/blob/1.3/nifty-core/src/main/resources/nifty-1.3.xsd

I've explained the issue with the XSD namespace in issue #106.

I think in the future it will be best to do what I said in that ticket: "The XSD that Nifty includes is the correct one for this Nifty version" So, I've now removed the old outdated XSD from the website to reduce further confusion.

In general a XSD namespace does not have to be an URL and even if it is, it's not required to be available online. So that should be fine - although I'm sure someone will complain soon :smile: 

Oh and XML will **not** be deprecated in 1.4 or in 2.0 at all so please continue to use it if you like XML.

PS: Nifty 2.0 XML might be different and not compatible with 1.x XML but that's something for 2014 or so ... :wink:
